src/Uno.UI/UI/Xaml/Controls/ContentPresenter/ContentPresenter.skia.cs

```diff
@@ -46,96 +47,127 @@ partial class ContentPresenter
 				throw new InvalidOperationException("ContentTemplateSelector cannot be set when the Content is a native element");
 			}
 
-			RegisterNativeHostSupport();
+			if (IsLoaded)
+			{
+				//If loaded, attach immediately. If not, don't attach since OnLoaded will attach later.
+				AttachNativeElement();
+			}
 		}
 		else if (IsNativeHost)
 		{
 			IsNativeHost = false;
-			UnregisterNativeHostSupport();
 		}
 	}
 
-	void RegisterNativeHostSupport()
+	private void ArrangeNativeElement()
 	{
-		if (IsNativeHost && XamlRoot is not null)
+		if (!IsNativeHost)
 		{
-			XamlRoot.InvalidateRender += UpdateNativeElementPosition;
-			_nativeHostRegistered = true;
+			// the ArrangeNativeElement call is queued on the dispatcher, so by the time we get here, the ContentPresenter
+			// might no longer be a NativeHost
+			return;
 		}
-	}
+		var arrangeRect = this.GetAbsoluteBoundsRect();
+		var ev = GetParentViewport().Effective;
 
-	void UnregisterNativeHostSupport()
-	{
-		if (_nativeHostRegistered)
+		Rect clippingBounds;
+		if (ev.IsEmpty)
+		{
+			clippingBounds = new Rect(0, 0, 0, 0);
+		}
+		else if (ev.IsInfinite)
+		{
+			clippingBounds = null;
+		}
+		else
 		{
-			_nativeHostRegistered = false;
-			XamlRoot.InvalidateRender -= UpdateNativeElementPosition;
+			var top = Math.Min(Math.Max(0, ev.Y), ActualHeight);
+			var height = Math.Max(0, Math.Min(ev.Height + ev.Y, ActualHeight - top));
+			var left = Math.Min(Math.Max(0, ev.X), ActualWidth);
+			var width = Math.Max(0, Math.Min(ev.Width + ev.X, ActualWidth - left));
+			clippingBounds = new Rect(left, top, width, height);
 		}
+
+		_nativeElementHostingExtension.Value!.ArrangeNativeElement(
+			Content,
+			arrangeRect,
+			clippingBounds);
 	}
 
-	partial void ArrangeNativeElement(Rect arrangeRect)
+	partial void AttachNativeElement()
 	{
-		if (IsNativeHost)
+#if DEBUG
```

Comment:
Debug.Assert is already done only in debug.

Reply:
The `#if DEBUG` is for the `_nativeElementAttached` assignment mainly.